neuralmonkey/decoders/sequence_labeler.py

```diff
@@ -142,6 +127,66 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
 
         sentences = dataset.maybe_get_series(self.data_id)
         if sentences is not None:
-            fd[self.target_tokens] = pad_batch(list(sentences))
+            fd[self.target_tokens] = pad_batch(
+                list(sentences), self.max_output_len, self.add_start_symbol,
+                self.add_end_symbol)
 
         return fd
+
+
+class EmbeddingsLabeler(SequenceLabeler):
+    """SequenceLabeler that uses an embedding matrix for output projection."""
+
+    # pylint: disable=too-many-arguments,too-many-locals
+    def __init__(self,
+                 name: str,
+                 encoders: List[TemporalStateful],
+                 embedded_sequence: EmbeddedSequence,
+                 data_id: str,
+                 max_output_len: int,
+                 hidden_dim: int = None,
+                 activation: Callable = tf.nn.relu,
+                 train_embeddings: bool = True,
+                 dropout_keep_prob: float = 1.0,
+                 add_start_symbol: bool = False,
+                 add_end_symbol: bool = False,
+                 reuse: ModelPart = None,
+                 save_checkpoint: str = None,
+                 load_checkpoint: str = None,
+                 initializers: InitializerSpecs = None) -> None:
+
+        check_argument_types()
+        SequenceLabeler.__init__(
+            self, name, encoders, embedded_sequence.vocabulary, data_id,
+            max_output_len, hidden_dim=hidden_dim, activation=activation,
+            dropout_keep_prob=dropout_keep_prob,
+            add_start_symbol=add_start_symbol, add_end_symbol=add_end_symbol,
+            reuse=reuse, save_checkpoint=save_checkpoint,
+            load_checkpoint=load_checkpoint, initializers=initializers)
+
+        self.embedded_sequence = embedded_sequence
+        self.train_embeddings = train_embeddings
+    # pylint: enable=too-many-arguments,too-many-locals
+
+    @tensor
+    def logits(self) -> tf.Tensor:
+        embeddings = self.embedded_sequence.embedding_matrix
+        if not self.train_embeddings:
+            embeddings = tf.stop_gradient(embeddings)
```

Comment:
Fakt muzes tohle udelat? Vzhledem k tomu, ze se jedna o fakticky posledni vrstvu, nestane se to, ze ti pri backpropu neprotece zadna informace do zbytku site (a tedy se nic nenaucis)?

Reply:
neproteče tam přes tuhle lokální proměnnou `embeddings`, ale proteče tam skrz `states`.

Reply:
Jasne, uz to vidim